prov/efa/src/efa_ep.c

```diff
@@ -699,24 +699,16 @@ int efa_ep_open(struct fid_domain *domain_fid, struct fi_info *user_info,
 		memcpy(ep->src_addr, user_info->src_addr, user_info->src_addrlen);
 	}
 
-	if (ep->domain->hmem_info[FI_HMEM_CUDA].initialized) {
-		/*
-		 * Set the default to required. NCCL plugin requires p2p, but
-		 * does not call setopt for this option in older NCCL plugin
-		 * versions.
-		 */
-		ep->hmem_p2p_opt = FI_HMEM_P2P_REQUIRED;
-	} else if (ep->domain->hmem_info[FI_HMEM_NEURON].initialized) {
-		/* Neuron requires p2p and supports no other modes. */
-		ep->hmem_p2p_opt = FI_HMEM_P2P_REQUIRED;
-	} else if (ep->domain->hmem_info[FI_HMEM_SYNAPSEAI].initialized) {
-		/* SynapseAI requires p2p and supports no other modes. */
-		ep->hmem_p2p_opt = FI_HMEM_P2P_REQUIRED;
-	} else {
-		/* no hmem devices, disable p2p */
-		ep->hmem_p2p_opt = FI_HMEM_P2P_DISABLED;
+	EFA_HMEM_IFACE_FOREACH_NON_SYSTEM(i) {
+		if (ep->domain->hmem_info[efa_hmem_ifaces[i]].initialized) {
+			ep->hmem_p2p_opt = ep->domain->hmem_info[efa_hmem_ifaces[i]].default_p2p_opt;
```

Comment:
I am not sure about this. What if there are 2 interfaces initialized? Today initialized just means libfabric is compiled with the support, and library has been loaded.

Reply:
It is possible for an instance to have both Neuron library and CUDA library installed. With this, neuron setting end up always overwrite CUDA setting.

Reply:
> I am not sure about this. What if there are 2 interfaces initialized? Today initialized just means libfabric is compiled with the support, and library has been loaded.

You and I spoke about this yesterday. I thought we agreed this was all we could really do given the way the `FI_HMEM` interfaces are modeled.

Also, this is no different than the previous logic. The previous logig was simply an unrolled version of this for loop.

> With this, neuron setting end up always overwrite CUDA setting.

You have it backwards. `FI_HMEM_CUDA` would be checked first (see enum definition) and break out of the loop before `FI_HMEM_NEURON` is checked at all.

Reply:
> Also, this is no different than the previous logic. The previous logic was simply an unrolled version of this for loop.

To demonstrate my point in pseudocode:

```c
    // Old
    if CUDA.initialized:
        p2p_opt = CUDA.default_p2p_opt
    else if NEURON.initialized:
        p2p_opt = NEURON.default_p2p_opt
    else if SYNAPSEAI.initialized:
        p2p_opt = SYNAPSEAI.default_p2p_opt
    else:
        p2p_opt = <default_p2p_opt>
```
is the same as
```c
    // New
    for iface in [CUDA, NEURON, SYNAPSEAI]:
        if iface.initialized:
            p2p_opt = iface.default_p2p_opt
            goto p2p_opt_set
    p2p_opt = <default_p2p_opt>

p2p_opt_set:
    ...
```

So while I would agree the default p2p option logic is flawed, it's simply a generic translation of what we currently have. So we can brainstorm how to approach this issue, but I don't think this should be a blocker for the PR as a whole since it virtually changes nothing wrt current behavior.

Reply:
We should refactor how `efa_ep->hmem_opt` is evaluated, but we do not need `hmem_info->default_hmem_opt`.

it can be derived from existing information in `hmem_info`

```
int default_hmem_opt()
{
ninited = 0;
EFA_HMEM_IFACE_FOREACH_NON_SYSTEM(i) {
  if (ep->domain->hmem_info[efa_hmem_ifaces[i]].initialized)  {
      if (ep->domain->hmem_info[efa_hmem_ifaces[i]].p2p_required_by_impl)
              return p2p_required;

    ninited +=1;
  }
  if (ninited == 0)
       return p2p_disabled;

  return p2p_prefered
}

int efa_ep_open()
{
....
efa_ep->hmem_opt = default_hmem_opt();
}
```